packages/block-editor/src/hooks/content-lock-ui.js

```diff
@@ -41,6 +47,89 @@ function StopEditingAsBlocksOnOutsideSelect( {
 	return null;
 }
 
+function ShufflePatternsToolbarItem( { clientId } ) {
+	// TODO: Probably worth to add this to blocks' selectors.
+	const getFlattenContentBlocks = useSelect( ( select ) => {
+		const contentBlockNames = select( blocksStore )
+			.getBlockTypes()
+			.filter(
+				( blockType ) =>
+					blockType.name !== 'core/list-item' &&
+					Object.values( blockType.attributes ).some(
+						( attribute ) =>
+							attribute.__experimentalRole === 'content'
+					)
+			)
+			.map( ( blockType ) => blockType.name );
+
+		return ( blocks ) =>
+			flattenBlocks( blocks ).filter( ( block ) =>
+				contentBlockNames.includes( block.name )
+			);
+	}, [] );
+	const { contentBlocks, patterns } = useSelect(
+		( select ) => {
+			const blocks =
+				select( blockEditorStore ).getBlocksByClientId( clientId );
+			const _contentBlocks = getFlattenContentBlocks( blocks );
+			const allPatterns =
+				select( blockEditorStore ).__experimentalGetAllowedPatterns();
+
+			return {
+				contentBlocks: _contentBlocks,
+				patterns: allPatterns
+					.filter( ( pattern ) => {
+						const patternContentBlocks = getFlattenContentBlocks(
+							pattern.blocks
+						);
+						return (
+							patternContentBlocks.length ===
+								_contentBlocks.length &&
+							_contentBlocks.every(
+								( block, index ) =>
+									block.name ===
+									patternContentBlocks[ index ].name
+							)
+						);
+					} )
+					.filter(
+						( pattern ) =>
+							! areBlocksAlike( blocks, pattern.blocks )
+					),
```

Comment:
Have you tried using `__experimentalGetPatternsByBlockTypes` or `__experimentalGetPatternTransformItems`?

Reply:
`__experimentalGetPatternsByBlockTypes` uses `blockTypes` in patterns, but unfortunately we don't have that information here.

`__experimentalGetPatternTransformItems` only handles blocks without inner blocks so we can't use that yet either.